code/go/internal/validator/semantic/validate_kibana_filter_present.go

```diff
@@ -0,0 +1,62 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package semantic
+
+import (
+	"errors"
+	"fmt"
+	"path"
+
+	"github.com/mitchellh/mapstructure"
+
+	ve "github.com/elastic/package-spec/v2/code/go/internal/errors"
+	"github.com/elastic/package-spec/v2/code/go/internal/fspath"
+	"github.com/elastic/package-spec/v2/code/go/internal/pkgpath"
+)
+
+// ValidateKibanaFilterPresent checks that all the dashboards included in a package
+// contain a filter, so only data related to its datasets is queried.
+func ValidateKibanaFilterPresent(fsys fspath.FS) ve.ValidationErrors {
+	var errs ve.ValidationErrors
+
+	filePaths := path.Join("kibana", "dashboard", "*.json")
+	dashboardFiles, err := pkgpath.Files(fsys, filePaths)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("error finding Kibana dashboard files: %w", err))
+		return errs
+	}
+	for _, file := range dashboardFiles {
+		err = checkDashboardHasFilter(file)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("file \"%s\" is invalid: expected filter in dashboard: %w", fsys.Path(file.Path()), err))
+		}
+	}
+
+	return errs
+}
+
+func checkDashboardHasFilter(file pkgpath.File) error {
+	searchJSON, err := file.Values("$.attributes.kibanaSavedObjectMeta.searchSourceJSON")
+	if err != nil {
+		return fmt.Errorf("unable to find search definition: %w", err)
+	}
+
+	var search struct {
+		Filter []interface{} `mapstructure:"filter"`
+		Query  struct {
+			Query string `mapstructure:"query"`
+		} `mapstructure:"query"`
+	}
+	err = mapstructure.Decode(searchJSON, &search)
+	if err != nil {
+		return fmt.Errorf("unable to decode search definition: %w", err)
+	}
+
+	if len(search.Filter) == 0 && len(search.Query.Query) == 0 {
+		return errors.New("no filter found")
+	}
```

Comment:
So at least a filter or a query should be present in the dashboard, is that right?

In an issue related, it was mentioned that it would be preferable to keep the search bar for the users: https://github.com/elastic/integrations/issues/4936#issuecomment-1387095936

But if we do not allow any query neither, maybe would the number of packages failing be greater ? Would it be too restrictive ?

Any recommendation about that @drewdaemon ?

Reply:
So I should allow only the filters and not the query? Sounds good to me. It will imply changes in more packages though.

Reply:
Changed to report a message also if there is a query, but not a filter. With different error messages.

Reply:
Kibana's UI presents both a query bar and filter pills, but they accomplish basically the same thing. A Kibana query is simply converted into a `filter` clause in the Elasticsearch DSL.

If we're adding this check to encourage more performant dashboards by prompting integration authors to scope down the underlying Elasticsearch queries, then I think that either a query or a filter would be acceptable.

Reply:
With current code we would only allow a filter, and the query is expected to be empty. So users can use the queries for quick filtering without risk of altering the provided filters. And this way we will also provide a more consistent experience between different dashboards.